accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,219 @@
+However with the pillar in database and loosely (no foreign key relation) connected to minions and groups, database trigger which can remove pillar entries when minion or group is removed:
+
+```sql
+CREATE OR REPLACE FUNCTION minion_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.minion_id;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE OR REPLACE FUNCTION group_removed() RETURNS TRIGGER AS $$
+BEGIN
+  DELETE FROM suseSaltPillars WHERE target = OLD.id::VARCHAR;
+  RETURN OLD;
+END $$ LANGUAGE PLPGSQL;
+
+CREATE TRIGGER minion_removed BEFORE DELETE ON suseMinionInfo FOR EACH ROW EXECUTE PROCEDURE minion_removed();
+
+CREATE TRIGGER group_removed BEFORE DELETE ON rhnServerGroup FOR EACH ROW EXECUTE PROCEDURE group_removed();
+```
+
+## Performance impact
+
+I did a small scale test with 100 minions and 100 pillar refresh repetitions with following results:
+
+Command:
+
+```bash
+time bash -c 'for i in `seq 1 100`;do salt "*" saltutil.pillar_refresh > /dev/null; done'
+```
+
+* both suma-minion.py and DB pillars active:
+
+```
+real    4m40,897s
+user    1m52,634s
+sys     0m10,143s
+```
+
+* only suma-minion.py pillar active:
+
+```
+real    4m37,656s
+user    1m53,964s
+sys     0m10,080s
+```
+
+
+* only DB pillar active:
+
+```
+real    4m39,619s
+user    1m54,091s
+sys     0m10,101s
+```
+
+
+I also did another small scale test with 1000 minions and 10 pillar refresh repetitions with following results:
+
+Command:
+```bash
+time bash -c 'for i in `seq 1 10`;do salt "*" saltutil.pillar_refresh > /dev/null; done'
+```
+
+* both DB pillar and `suma_minion.py` pillars active:
+
+```
+real    3m17,522s
+user    0m39,096s
+sys     0m2,260s
+```
+
+* when only `suma_minion.py` pillar was active:
+
+```
+real    3m22,705s
+user    0m40,239s
+sys     0m2,244s
+```
+
+* when only DB pillar was active:
+
+```
+real    3m21,272s
+user    0m39,661s
+sys     0m2,320s
+```
+
+In conclusion I do not see very notable impact in this isolated scenario.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we **not** do this?
+
+* what other parts of the product will be affected?
+  
+  Salt integration and thus basically whole product is affected.
+  
+* will the solution be hard to maintain in the future?
+  
+  I expect easier maintaining then with file generators.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Use single pillar row for each minion and without `category` column.
+
+This was the original idea how to implement this, however during prototype implementation having separate rows for separate pillar data made implementation better match how it is done now. I am not opposed using single row per minion, but I also do not see disadvantages for using multiple rows per minion.
```

Comment:
IMHO One disadvantage of this alternative approach it will be harder to update the json data in the database